uaiowebsockets/protocol.py

```diff
@@ -0,0 +1,248 @@
+"""
+Websockets protocol
+"""
+
+import logging
+import ure as re
+import ustruct as struct
+import urandom as random
+import usocket as socket
+from ucollections import namedtuple
+
+LOGGER = logging.getLogger(__name__)
+
+# Opcodes
+OP_CONT = const(0x0)
+OP_TEXT = const(0x1)
+OP_BYTES = const(0x2)
+OP_CLOSE = const(0x8)
+OP_PING = const(0x9)
+OP_PONG = const(0xa)
+
+# Close codes
+CLOSE_OK = const(1000)
+CLOSE_GOING_AWAY = const(1001)
+CLOSE_PROTOCOL_ERROR = const(1002)
+CLOSE_DATA_NOT_SUPPORTED = const(1003)
+CLOSE_BAD_DATA = const(1007)
+CLOSE_POLICY_VIOLATION = const(1008)
+CLOSE_TOO_BIG = const(1009)
+CLOSE_MISSING_EXTN = const(1010)
+CLOSE_BAD_CONDITION = const(1011)
+
+URL_RE = re.compile(r'(wss|ws)://([A-Za-z0-9-\.]+)(?:\:([0-9]+))?(/.+)?')
+URI = namedtuple('URI', ('protocol', 'hostname', 'port', 'path'))
+
+class NoDataException(Exception):
+    pass
+
+class ConnectionClosed(Exception):
+    pass
+
+def urlparse(uri):
+    """Parse ws:// URLs"""
+    match = URL_RE.match(uri)
+    if match:
+        protocol = match.group(1)
+        host = match.group(2)
+        port = match.group(3)
+        path = match.group(4)
+
+        if protocol == 'wss':
+            if port is None:
+                port = 443
+        elif protocol == 'ws':
+            if port is None:
+                port = 80
+        else:
+            raise ValueError('Scheme {} is invalid'.format(protocol))
+
+        return URI(protocol, host, int(port), path)
+
+
+class Websocket:
+    """
+    Basis of the Websocket protocol.
+
+    This can probably be replaced with the C-based websocket module, but
+    this one currently supports more options.
+    """
+    is_client = False
+
+    def __init__(self, sock, sreader, swriter):
+        self.sock = sock
+        self.sreader = sreader
+        self.swriter = swriter
+        self.open = True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc, tb):
```

Comment:
This probably now needs to be `aexit`, so you can wrap up the destruction of your websocket automatically.

Reply:
`aenter` too then, yes?

Reply:
Changed both and added another example